### ISSUE TITLE
Profilers do not clean up after the avocado run finishes. 

### ISSUE THREAD

**Issue by Venutiwa** [avocado-framework/avocado#2135]
Profilers added in /etc/avocado/sysinfo/profilers keep running even after the avocado run finishes, filling up the 100% of the disk space. profilers should get cleaned/killed once the test completes.
sample ps -ef output after avocado-setup.py exit:
root 170073 2 0 02:40 ? 00:00:00 [kworker/0:2]
root 170309 1 0 04:08 pts/0 00:00:02 iostat -t -x 5
root 170316 1 0 04:08 pts/0 00:00:00 vmstat -t 5
root 170319 1 1 04:08 pts/0 00:00:15 top -b -d 5 -H
root 170329 1 0 04:08 pts/0 00:00:01 iostat -t 5
root 170341 1 0 04:08 pts/0 00:00:03 mpstat -P ALL 5
root 170646 81014 0 04:08 ? 00:00:00 sshd: root@pts/5
One has to manually kill the profilers after each run. It is bit severe as the test unattended would pull the SUT down due to / getting filled.

**Comment by apahim** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [[BUG] Profilers do not clean up after the avocado run finishes.](https://trello.com/c/QUhYcX5A/1091-bug-profilers-do-not-clean-up-after-the-avocado-run-finishes)

**Comment by apahim** [comment]
Hi @Venutiwa, I'm not able to reproduce this issue. In my system, the profilers are terminated at the end of the Job. Could you please provide the `job.log` of your execution?

**Comment by Venutiwa** [comment]
Attaching the job.log for host system test run.

(See attached file: job.log)

Thanks and Regards,
Venu Tiwari
Powersystems Performance
Systems and Technology Group
India Software Lab,IBM
EGL, Bangalore
9535633355

Happiness can be found in the darkest of times, if one only remembers to
turn on the light.
-Albus Percival Wulfric Brian Dumbledore

**Comment by Venutiwa** [comment]
[job.zip](https://github.com/avocado-framework/avocado/files/1241097/job.zip)
 Attached job.log for the host system test run.

**Comment by apahim** [comment]
Hmm, job.log is no help. I thought we would find there some messages related to the profilers not being finished.
Could you please send me the result of `ps fxa -o ppid,pid,stat,time,command`, execute during the test and executed agian after the test?

**Comment by Venutiwa** [comment]
[profilers_bug.zip](https://github.com/avocado-framework/avocado/files/1241428/profilers_bug.zip)
 attached required data along with fresh job.log

**Comment by apahim** [comment]
Ok, I think we are close :)
Can I see your `/etc/avocado/sysinfo/profilers`?

**Comment by Venutiwa** [comment]
[profilers.zip](https://github.com/avocado-framework/avocado/files/1241537/profilers.zip)

**Comment by apahim** [comment]
Ok, even though I cannot reproduce, I have a good idea on what´s happening. I see your system is creating the calling `/bin/sh` to execute the commands, which are executed in a subprocess. When we terminate the profiler we are in fact terminating the `/bin/sh` and not the subprocess created by it.
To prove my theory, I'd like to request you to make a little change in the `sysinfo.py` source code:

```diff
diff --git a/avocado/core/sysinfo.py b/avocado/core/sysinfo.py
index 3e145ba5..56e5826d 100644
--- a/avocado/core/sysinfo.py
+++ b/avocado/core/sysinfo.py
@@ -16,6 +16,7 @@ import gzip
 import json
 import logging
 import os
+import shlex
 import shutil
 import time
 import threading
@@ -222,8 +223,8 @@ class Daemon(Command):
         logf_path = os.path.join(logdir, self.logf)
         stdin = open(os.devnull, "r")
         stdout = open(logf_path, "w")
-        self.pipe = subprocess.Popen(self.cmd, stdin=stdin, stdout=stdout,
-                                     stderr=subprocess.STDOUT, shell=True, env=env)
+        self.pipe = subprocess.Popen(shlex.split(self.cmd), stdin=stdin, stdout=stdout,
+                                     stderr=subprocess.STDOUT, shell=False, env=env)

     def stop(self):
         """
```
Notice that the `subprocess.Popen` call  to be changed is the one present in the class `Daemon`.

Depending on how you've installed Avocado, this file can be in different locations. If you're using it from dnf/rpm, location is `/usr/lib/python2.7/site-packages/avocado/core/sysinfo.py`.

After that change, you can run Avocado again and see what happens. Looking forward to see the results.

**Comment by Venutiwa** [comment]
Hi,
I made changes as per your directions, but still observing the same behavior.
Here are the latest logs:
[profilers_bugAug23.zip](https://github.com/avocado-framework/avocado/files/1244501/profilers_bugAug23.zip)

**Comment by apahim** [comment]
Ok, thank you. You can restore the file.
What troubles me is this behavior:
```bash
185332 185366 S+   00:00:00  |                   \_ /bin/sh -c iostat -t -x 5
185366 185367 S+   00:00:00  |                   |   \_ iostat -t -x 5
```
This extra parent process `/bin/sh -c iostat -t -x 5` cannot be observed in our systems and my guess is that it's the source of the problem.

**Comment by apahim** [comment]
Well, I'm having a hard time trying to reproduce the issue. Could you describe the steps to reproduce it? What distro/version and so on?

**Comment by Venutiwa** [comment]
I am triggering avocado-setup.py script to run fio tests:
something like this:
python avocado-setup.py --run-suite guest_fio --guest-os Ubuntu.16.04.3.ppc64le --vt qemu --no-download
python avocado-setup.py --run-suite host_fio --no-download

Following are the system details:
Linux 4.4.0-62-generic #83-Ubuntu SMP Wed Jan 18 14:09:19 UTC 2017 ppc64le ppc64le ppc64le GNU/Linux
NAME="Ubuntu"
VERSION="16.04.2 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.2 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial

**Comment by ldoktor** [comment]
Well I'm also unable to reproduce it, but from what I see your system is really protective in terms of `subprocess` processes. Perhaps we should be more aggressive and use:

```
diff --git a/avocado/core/sysinfo.py b/avocado/core/sysinfo.py
index 3e145ba..adb8c17 100644
--- a/avocado/core/sysinfo.py
+++ b/avocado/core/sysinfo.py
@@ -231,7 +231,7 @@ class Daemon(Command):
         """
         retcode = self.pipe.poll()
         if retcode is None:
-            self.pipe.terminate()
+            process.kill_process_tree(self.pipe.pid)
             retcode = self.pipe.wait()
         else:
             log.error("Daemon process '%s' (pid %d) terminated abnormally (code %d)",
```

Does that solve your issue?

**Comment by Venutiwa** [comment]
No, sorry but it does not.

**Comment by ldoktor** [comment]
That is odd. Do you use job or test profilers? Because I noticed we might have bug in the test profilers, can you try this patch and log here the output of some sample execution (like `avocado run passtest.py`)? If there are any lines with the `end_test_hook daemon` then we have (probably) a winner...

```diff
diff --git a/avocado/core/sysinfo.py b/avocado/core/sysinfo.py
index 3e145ba..dd8dc4f 100644
--- a/avocado/core/sysinfo.py
+++ b/avocado/core/sysinfo.py
@@ -34,6 +34,7 @@ from ..utils import path as utils_path
 from ..utils import wait
 
 log = logging.getLogger("avocado.sysinfo")
+LOG = logging.getLogger("avocado.app")
 
 
 class Collectible(object):
@@ -572,6 +573,7 @@ class SysInfo(object):
         Logging hook called whenever a job starts.
         """
         for log in self.start_job_collectibles:
+            LOG.critical("start_job_hook: %s", log)
             if isinstance(log, Daemon):  # log daemons in profile directory
                 log.run(self.profile_dir)
             else:
@@ -585,10 +587,12 @@ class SysInfo(object):
         Logging hook called whenever a job finishes.
         """
         for log in self.end_job_collectibles:
+            LOG.critical("end_job_hook: %s", log)
             log.run(self.post_dir)
         # Stop daemon(s) started previously
         for log in self.start_job_collectibles:
             if isinstance(log, Daemon):
+                LOG.critical("end_job_hook daemon: %s", log)
                 log.stop()
 
         if self.log_packages:
@@ -599,7 +603,14 @@ class SysInfo(object):
         Logging hook called before a test starts.
         """
         for log in self.start_test_collectibles:
+            LOG.critical("start_test_hook: %s", log)
             log.run(self.pre_dir)
+        # Stop daemon(s) started previously
+        for log in self.start_test_collectibles:
+            if isinstance(log, Daemon):
+                LOG.critical("end_test_hook daemon: %s", log)
+                log.stop()
+
 
         if self.log_packages:
             self._log_installed_packages(self.pre_dir)
@@ -609,6 +620,7 @@ class SysInfo(object):
         Logging hook called after a test finishes.
         """
         for log in self.end_test_collectibles:
+            LOG.critical("end_test_hook: %s", log)
             log.run(self.post_dir)
 
         if self.log_packages:
```

**Comment by Venutiwa** [comment]
[profiler_issue_sept12.zip](https://github.com/avocado-framework/avocado/files/1294894/profiler_issue_sept12.zip)

Sorry for the delay,
Made the above changes and attached the job.log for the test, although I do not see the "end_test_hook daemon" in the logs.

Question:Do you use job or test profilers?
Answer: I think these are job profilers not for individual tests, as we do not see any any test profilers running. I have raised an issue regarding the same: 
https://github.com/avocado-framework/avocado/issues/2169

**Comment by ldoktor** [comment]
Hello @Venutiwa, I used the `avocado.app` logger for debugging messages, so they were present in to human output and not in the log itself. Could you please also include the UI output? Alternatively just replace the `LOG = logging.getLogger("avocado.app")` with `LOG = logging.getLogger("avocado.sysinfo")` which should store the messages in `job.log` and include it here.

Yes, the per-test-profilers are not yet supported, even though the infrastructure partially exists. I asked because I was curious, whether you used that (undescribed) infrastructure which might (and actually contains) errors.

Anyway I'm really curious about the debug messages, hopefully it'll give us better insight.

**Comment by Venutiwa** [comment]
[profilers_bug_sept13.zip](https://github.com/avocado-framework/avocado/files/1298432/profilers_bug_sept13.zip)

Hi @ldoktor  attached the latest logs with LOG = logging.getLogger("avocado.sysinfo") .
When you say UI output, do you mean console output?I have added them as well.

Again I did not see "end_test_hook daemon" in the job.log.

**Comment by ldoktor** [comment]
Well there are not even the `start_job_hook` log entries which means the patched code was not executed. Are you sure you applied the patch correctly and (more importantly) you ran `make install` or `make develop` afterwards to propagate the change? To make sure please change the version number to something else to see whether that particular version is the executed one:

```diff
diff --git a/VERSION b/VERSION
index 7eb27e3..4ee738e 100644
--- a/VERSION
+++ b/VERSION
@@ -1 +1 @@
-53.0
+88.8
```

**Comment by apahim** [comment]
@Venutiwa, considering we don't have a reproducer and the debug process is quite slow without one, how much trouble would that be for you to provide us access to one affected machine? If that's feasible, please reach us via e-mail to arrange the access. Thanks.

**Comment by Venutiwa** [comment]
Hey @apahim  @ldoktor, 
Here is the current status. Apparently i was applying all the changes in sysinfo.py located in the wrong location. Located the correct file and started applying the changes from first code changes:
1.) -        self.pipe = subprocess.Popen(self.cmd, stdin=stdin, stdout=stdout,
-                                     stderr=subprocess.STDOUT, shell=True, env=env)
+        self.pipe = subprocess.Popen(shlex.split(self.cmd), stdin=stdin, stdout=stdout,
+                                     stderr=subprocess.STDOUT, shell=False, env=env)

Received the following error:
1st patch:

root@perf2822:/var/lib/libvirt/images/workspace/runAvocadoFVTTest/avocado-fvt-wrapper# python avocado-setup.py --run-suite host_fio_seq_read --no-download
03:55:19 INFO    : Check for environment
03:55:19 INFO    : Creating temporary mux dir
03:55:19 INFO    : Check if bootstrap required
03:55:21 DEBUG   : Running 'which avocado'
03:55:21 INFO    :
03:55:21 INFO    : Running Host Tests Suite fio_seq_read_fiotest_fio_seq_read
03:55:21 DEBUG   : Running: /usr/local/bin/avocado run avocado-misc-tests/io/disk/fiotest.py -m /var/lib/libvirt/images/workspace/runAvocadoFVTTest/avocado-fvt-wrapper/tests/avocado-misc-tests/io/disk/fiotest.py.data/fio_seq_read.yaml --force-job-id aaad23446959fdef7ee8e9740caea91d8970a5b0  --job-results-dir /var/lib/libvirt/images/workspace/runAvocadoFVTTest/avocado-fvt-wrapper/results
JOB ID     : aaad23446959fdef7ee8e9740caea91d8970a5b0
JOB LOG    : /var/lib/libvirt/images/workspace/runAvocadoFVTTest/avocado-fvt-wrapper/results/job-2017-09-14T03.55-aaad234/job.log

Avocado crashed: NameError: global name 'shlex' is not defined
Traceback (most recent call last):

  File "/usr/local/lib/python2.7/dist-packages/avocado_framework-53.0-py2.7.egg/avocado/core/job.py", line 448, in run_tests
    execution_order)

  File "/usr/local/lib/python2.7/dist-packages/avocado_framework-53.0-py2.7.egg/avocado/core/runner.py", line 560, in run_suite
    self.job.sysinfo.start_job_hook()

  File "/usr/local/lib/python2.7/dist-packages/avocado_framework-53.0-py2.7.egg/avocado/core/sysinfo.py", line 576, in start_job_hook
    log.run(self.profile_dir)

  File "/usr/local/lib/python2.7/dist-packages/avocado_framework-53.0-py2.7.egg/avocado/core/sysinfo.py", line 226, in run
    self.pipe = subprocess.Popen(shlex.split(self.cmd), stdin=stdin, stdout=stdout, stderr=subprocess.STDOUT, shell=False, env=env)

NameError: global name 'shlex' is not defined

Please include the traceback info and command line used on your bug report
Report bugs visiting https://github.com/avocado-framework/avocado/issues/new
JOB HTML   : /var/lib/libvirt/images/workspace/runAvocadoFVTTest/avocado-fvt-wrapper/results/job-2017-09-14T03.55-aaad234/results.html
03:55:23 INFO    :
03:55:23 INFO    : Summary of test results can be found below:
TestSuite                 Testrun    ResultLink                                                                            Summary
host_fio_seq_read_fiotest_fio_seq_read Run        /var/lib/libvirt/images/workspace/runAvocadoFVTTest/avocado-fvt-wrapper/results/job-2017-09-14T03.55-aaad234/job.log Successfully executed
03:55:23 INFO    : Removing temporary mux dir

2.)-            self.pipe.terminate()
+            process.kill_process_tree(self.pipe.pid)
Applied the above code changes and issue is not re-produced. 
Looks like the 2nd code fix works.

**Comment by apahim** [comment]
Oh, cool. From the 1st patch, you missed the `import shlex`. Just to make sure the issue is/is not  related to the `shell=True`, could you include the `import shlex` and try again?

Anyway, with the second patch working, we can provide that a fix since we really should kill the process tree (not only the first child).

Thanks!

**Comment by Venutiwa** [comment]
Yes, after adding "import shlex" , 1st patch works too.  I do not see profilers running after the job exits.

Sorry for the delay. Thanks for all the help, appreciate it.

**Comment by apahim** [comment]
Fixed by #2207
